chapter/4/dist-langs.md

```diff
@@ -4,7 +4,515 @@ title:  "Distributed Programming Languages"
+When a guardian crashes and is restored, this serialized state is used to reconstruct guardians.
+Like in Emerald, internal data members may not be accessed directly, but rather through handlers.
+Guardians are interacted with through methods known as *handlers*.
+When a handler is called, a new process is created to handle the operation.
+Additionally, guardians may contain a background process for performing continual work.
+
+Argus encapsulates handler calls in what it calls *actions*.
+Actions are designed to solve the problems of consistency, synchronization, and fault tolerance.
+To accomplish this, actions are serializable as well as total.
+Being serializable means that no actions interfere with one another.
+For example, a read operation that spans multiple guardians either "sees" the complete effect of a simultaneous write operation, or it sees nothing.
+Being total means that write operations that span multiple guardians either fully complete or fully fail.
+This is accomplished by a two-phase commit protocol that serializes the state of *all* guardians involved in an action, or discards partial state changes.
+
+#### Erlang (2000)
+
+Erlang is a distributed language which combines functional programming with message passing.
+Units of distribution in Erlang are processes.
+These processes may be co-located on the same node or distributed amongst a cluster.
+
+Processes in Erlang communicate by message passing.
+Specifically, a process can send an asynchronous message to another process' *mailbox*.
+At some future time, the receiving process may enter a *receive* clause, which searches the mailbox for the first message that matches one of a set of patterns.
+The branch of code that is executed in response to a message is dependent on the pattern that is matched.
+
+In general, an application written in Erlang is separated into two broad components: *workers* and *monitors*.
+Workers are responsible for application logic.
+Erlang offers a special function `link(Pid)` which allows one process to monitor another.
+If the process indicated by `Pid` fails, the monitor process will be notified and is expected to handle the error.
+Worker processes are "linked" by monitor processes which implement the fault-tolerance logic of the application.
+
+Erlang, first implemented in Prolog, has the features and styles of a functional programming language.
+Variables in Erlang are immutable; once assigned a value, they cannot be changed.
+Because of this, loops in Erlang are written as tail recursive function calls.
+Although this at first seems like a flawed practice (to traditional procedural programmers), tail recursive calls do not grow the current stack frame, but instead replace it.
+It is worth noting that stack frame growth is still possible, but if the recursive call is "alone" (the result of the inner function is the result of the outer function), the stack will not grow.
```

Comment:
Does this detail really matter?  Seems like a minor Erlang detail where this section doesn't discuss anything specific to Distributed Erlang at all.